### PR TITLE
Fix issue #210

### LINKED ISSUE #210: Loki does not provide the ingress address over the prometheus_scrape relation
### Bug Description

Loki sends it's internal address to prometheus rather than it's ingress address.

### To Reproduce

* Deploy Loki, Prometheus, and Traefik
* Relate Loki -> Traefik and Loki -> Prometheus
* Check targets page in prometheus UI

### Environment

Ubuntu 22.04
Juju latest/stable

### Relevant log output

```shell
-
```

### Additional context

image

### PR DESCRIPTION
**Pull request by Abuelodelanada**
## Issue

This PR fixes #210 

## Solution

Now we update scrape job using the new public method in `MetricsEndpointProvider`

## Context

This PR relies in [this Prometheus' PR](https://github.com/canonical/prometheus-k8s-operator/pull/397)

## Testing Instructions

Follow the instructions in [this Prometheus' PR](https://github.com/canonical/prometheus-k8s-operator/pull/397)

### CODE REVIEW

**Review by dstathis — APPROVED** [review]

**Review by rbarry82 — APPROVED** [review]

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 1]
```diff
@@ -936,7 +936,7 @@ def _resolve_dir_against_charm_path(charm: CharmBase, *path_elements: str) -> st
 class NoRelationWithInterfaceFoundError(Exception):
     """No relations with the given interface are found in the charm meta."""
 
-    def __init__(self, charm: CharmBase, relation_interface: str = None):
+    def __init__(self, charm: CharmBase, relation_interface: str = ""):
```
Comment (sed-i) [inline lib/charms/loki_k8s/v0/loki_push_api.py:939]: These mypy errors can be ignored with setting `implicit-optional` to False.

File: lib/charms/loki_k8s/v0/loki_push_api.py [hunk 2]
```diff
@@ -1676,7 +1676,7 @@ class LogProxyConsumer(ConsumerBase):
     def __init__(
         self,
         charm,
-        log_files: list = None,
+        log_files: list = [],
```
Comment (sed-i) [inline lib/charms/loki_k8s/v0/loki_push_api.py:1679]: Mutable default arg :(

File: src/charm.py [hunk 3]
```diff
@@ -160,6 +159,7 @@ def _on_ingress_changed(self, _):
         self._configure()
         self.loki_provider.update_endpoint(url=self._external_url)
         self.grafana_source_provider.update_source(source_url=self._external_url)
+        self.metrics_provider.update_scrape_job_spec(self.scrape_jobs)
```
Comment (sed-i) [inline src/charm.py:162]: Have you considered
- pass `self.scrape_jobs` as a callable to MetricsEndpointProvider; **and**
- pass `self._on_ingress_changed` as a refresh event to `self.metrics_provider`?

Iiuc, in that case we wouldn't need `update_scrape_job_spec` and users won't need to remember to call it.

File: lib/charms/loki_k8s/v0/loki_push_api.py (hunk 2 above)
Comment (Abuelodelanada) [reply lib/charms/loki_k8s/v0/loki_push_api.py:1679]: damm!

File: src/charm.py (hunk 3 above)
Comment (Abuelodelanada) [reply src/charm.py:162]: Yes, none of the above examples really worked well. Take a look at this: https://github.com/canonical/prometheus-k8s-operator/pull/397#issue-1446072828

**Pull request merged**